### ISSUE TITLE
Implement Statistical Arbitrage with Cointegration

### ISSUE THREAD

**Issue by Mathews-25** [AgesEmpire/StellarSwipe-Contract#109]
Context:
Statistical arbitrage identifies mean-reverting relationships between multiple assets using advanced statistical methods like cointegration, going beyond simple pairs trading.
Objective:
Build statistical arbitrage system using cointegration analysis for multi-asset portfolios.
Scope:

Test asset groups for cointegration
Build cointegrated portfolios
Calculate optimal hedge ratios via regression
Trade mean reversion in portfolio residuals
Support 3+ asset stat arb baskets

Implementation Guidelines:

Key Files: /contracts/auto_trade/src/strategies/stat_arb.rs
Statistical arbitrage structure:

rust  struct StatArbStrategy {
      user: Address,
      asset_basket: Vec<AssetPair>,
      lookback_period_days: u32,
      cointegration_threshold: i128, // ADF test statistic threshold
      hedge_ratios: Vec<i128>, // Optimal weights for each asset
      entry_z_score: i128,
      exit_z_score: i128,
      rebalance_frequency_hours: u32,
      active_portfolio: Option<StatArbPortfolio>,
      residual_history: Vec<i128>,
  }
  
  struct StatArbPortfolio {
      portfolio_id: u64,
      asset_positions: Map<AssetPair, i128>, // Amount in each asset
      entry_residual: i128,
      entry_z_score: i128,
      entry_time: u64,
      total_value: i128,
  }
  
  struct CointegrationTest {
      asset_group: Vec<AssetPair>,
      is_cointegrated: bool,
      adf_statistic: i128,
      p_value: u32,
      hedge_ratios: Vec<i128>,
      half_life: u64, // Mean reversion half-life in seconds
  }
Cointegration Testing:
rustfn test_cointegration(
    assets: Vec<AssetPair>,
    lookback_days: u32
) -> Result<CointegrationTest, Error> {
    require!(assets.len() >= 2 && assets.len() <= 5, "Support 2-5 assets");
    
    let lookback_seconds = lookback_days as u64 * 86400;
    
    // Collect price series for all assets
    let mut price_series = Vec::new();
    for asset in &assets {
        let prices = get_historical_prices(*asset, lookback_seconds)?;
        price_series.push(prices);
    }
    
    // Ensure all series same length
    let min_length = price_series.iter().map(|p| p.len()).min().unwrap();
    for series in &mut price_series {
        series.truncate(min_length);
    }
    
    // Find optimal hedge ratios using OLS regression
    let hedge_ratios = calculate_hedge_ratios_ols(&price_series)?;
    
    // Construct portfolio (linear combination)
    let portfolio_values = construct_portfolio(&price_series, &hedge_ratios)?;
    
    // Perform Augmented Dickey-Fuller test on portfolio residuals
    let adf_result = augmented_dickey_fuller_test(&portfolio_values)?;
    
    // Calculate half-life of mean reversion
    let half_life = calculate_mean_reversion_halflife(&portfolio_values)?;
    
    // Determine if cointegrated (ADF statistic < critical value)
    // Critical value at 5% significance: approximately -3.4 (scaled by 10000)
    let is_cointegrated = adf_result.statistic < -34000;
    
    Ok(CointegrationTest {
        asset_group: assets,
        is_cointegrated,
        adf_statistic: adf_result.statistic,
        p_value: adf_result.p_value,
        hedge_ratios,
        half_life,
    })
}

fn calculate_hedge_ratios_ols(
    price_series: &Vec<Vec<i128>>
) -> Result<Vec<i128>, Error> {
    // Use first asset as dependent variable, others as independent
    let y = &price_series[0];
    let x_matrix: Vec<Vec<i128>> = price_series[1..].to_vec();
    
    // OLS: β = (X'X)^-1 X'y
    let mut coefficients = vec![PRECISION]; // First asset gets weight of 1
    
    for i in 0..x_matrix.len() {
        let x = &x_matrix[i];
        
        // Simple bivariate regression for each independent variable
        let beta = calculate_regression_coefficient(y, x)?;
        coefficients.push(beta);
    }
    
    Ok(coefficients)
}

fn calculate_regression_coefficient(
    y: &[i128],
    x: &[i128]
) -> Result<i128, Error> {
    let n = y.len() as i128;
    
    let sum_x: i128 = x.iter().sum();
    let sum_y: i128 = y.iter().sum();
    let sum_xy: i128 = x.iter().zip(y.iter()).map(|(xi, yi)| (xi * yi) / n).sum();
    let sum_xx: i128 = x.iter().map(|xi| (xi * xi) / n).sum();
    
    let mean_x = sum_x / n;
    let mean_y = sum_y / n;
    
    // β = Cov(X,Y) / Var(X)
    let covariance = sum_xy - (mean_x * mean_y);
    let variance = sum_xx - (mean_x * mean_x);
    
    if variance == 0 {
        return Ok(0);
    }
    
    Ok((covariance * PRECISION) / variance)
}

fn construct_portfolio(
    price_series: &Vec<Vec<i128>>,
    hedge_ratios: &[i128]
) -> Result<Vec<i128>, Error> {
    let length = price_series[0].len();
    let mut portfolio_values = vec![0i128; length];
    
    for i in 0..length {
        let mut value = 0i128;
        for (j, series) in price_series.iter().enumerate() {
            value += (series[i] * hedge_ratios[j]) / PRECISION;
        }
        portfolio_values[i] = value;
    }
    
    Ok(portfolio_values)
}
Augmented Dickey-Fuller Test:
ruststruct ADFResult {
    statistic: i128,
    p_value: u32,
}

fn augmented_dickey_fuller_test(series: &[i128]) -> Result<ADFResult, Error> {
    // Simplified ADF test
    // Full implementation would include lag selection and critical value tables
    
    // Calculate first differences
    let mut differences = Vec::new();
    for i in 1..series.len() {
        differences.push(series[i] - series[i-1]);
    }
    
    // Regress Δy_t on y_(t-1)
    let y_lagged = &series[0..series.len()-1];
    let delta_y = &differences;
    
    let coefficient = calculate_regression_coefficient(delta_y, y_lagged)?;
    
    // Calculate t-statistic
    let residuals = calculate_residuals(delta_y, y_lagged, coefficient)?;
    let std_error = calculate_standard_error(&residuals)?;
    
    let t_statistic = if std_error > 0 {
        (coefficient * 10000) / std_error
    } else {
        0
    };
    
    // Estimate p-value from t-statistic
    // In practice, use lookup table or MacKinnon approximation
    let p_value = estimate_adf_pvalue(t_statistic)?;
    
    Ok(ADFResult {
        statistic: t_statistic,
        p_value,
    })
}

fn calculate_mean_reversion_halflife(series: &[i128]) -> Result<u64, Error> {
    // Calculate AR(1) coefficient
    let lagged = &series[0..series.len()-1];
    let current = &series[1..];
    
    let ar_coefficient = calculate_regression_coefficient(current, lagged)?;
    
    // Half-life = -ln(2) / ln(ρ) where ρ is AR(1) coefficient
    // Approximation: HL ≈ 0.693 / (1 - ρ)
    let rho = ar_coefficient as f64 / PRECISION as f64;
    
    if rho >= 1.0 {
        return Ok(u64::MAX); // No mean reversion
    }
    
    let half_life_periods = 0.693 / (1.0 - rho);
    
    // Convert to seconds (assuming daily data)
    Ok((half_life_periods * 86400.0) as u64)
}
Portfolio Construction & Trading:
rustfn check_stat_arb_signal(
    strategy_id: u64
) -> Result<Option<StatArbSignal>, Error> {
    let strategy = get_stat_arb_strategy(strategy_id)?;
    
    if strategy.active_portfolio.is_some() {
        return Ok(None);
    }
    
    // Test for cointegration
    let coint_test = test_cointegration(
        strategy.asset_basket.clone(),
        strategy.lookback_period_days
    )?;
    
    if !coint_test.is_cointegrated {
        return Err(Error::AssetsNotCointegrated);
    }
    
    // Calculate current portfolio value
    let mut current_prices = Vec::new();
    for asset in &strategy.asset_basket {
        current_prices.push(get_current_price(*asset)?);
    }
    
    let current_portfolio_value = calculate_portfolio_value(
        &current_prices,
        &coint_test.hedge_ratios
    )?;
    
    // Calculate z-score of current value vs historical mean
    let historical_mean = strategy.residual_history.iter().sum::<i128>() / 
                         strategy.residual_history.len() as i128;
    let std_dev = calculate_std_dev(&strategy.residual_history)?;
    
    let z_score = if std_dev > 0 {
        ((current_portfolio_value - historical_mean) * 10000) / std_dev
    } else {
        0
    };
    
    // Check if z-score exceeds entry threshold
    if z_score.abs() >= strategy.entry_z_score {
        let signal = StatArbSignal {
            hedge_ratios: coint_test.hedge_ratios,
            entry_residual: current_portfolio_value,
            z_score,
            expected_mean: historical_mean,
            half_life: coint_test.half_life,
            direction: if z_score > 0 {
                PortfolioDirection::Short // Portfolio overvalued, short it
            } else {
                PortfolioDirection::Long // Portfolio undervalued, long it
            },
        };
        
        Ok(Some(signal))
    } else {
        Ok(None)
    }
}

fn execute_stat_arb_trade(
    strategy_id: u64,
    signal: StatArbSignal
) -> Result<u64, Error> {
    let strategy = get_stat_arb_strategy_mut(strategy_id)?;
    
    let portfolio_value = get_portfolio_value(strategy.user)?;
    let total_allocation = (portfolio_value * 20) / 100; // 20% of portfolio
    
    // Calculate position sizes for each asset based on hedge ratios
    let mut asset_positions = Map::new();
    let total_hedge_ratio: i128 = signal.hedge_ratios.iter().sum();
    
    for (i, asset) in strategy.asset_basket.iter().enumerate() {
        let weight = signal.hedge_ratios[i];
        let asset_allocation = (total_allocation * weight) / total_hedge_ratio;
        
        // Adjust for long/short direction
        let position_amount = match signal.direction {
            PortfolioDirection::Long => asset_allocation,
            PortfolioDirection::Short => -asset_allocation,
        };
        
        // Execute trade
        execute_trade(
            strategy.user,
            *asset,
            position_amount.abs(),
            if position_amount > 0 {
                OrderType::Market // Buy
            } else {
                OrderType::Market // Sell
            }
        )?;
        
        asset_positions.insert(*asset, position_amount);
    }
    
    let portfolio_id = get_next_portfolio_id();
    
    let portfolio = StatArbPortfolio {
        portfolio_id,
        asset_positions,
        entry_residual: signal.entry_residual,
        entry_z_score: signal.z_score,
        entry_time: current_time(),
        total_value: total_allocation,
    };
    
    strategy.active_portfolio = Some(portfolio);
    
    emit_event(StatArbPortfolioCreated {
        strategy_id,
        portfolio_id,
        num_assets: strategy.asset_basket.len() as u32,
        entry_z_score: signal.z_score,
    });
    
    Ok(portfolio_id)
}
Rebalancing:
rustfn rebalance_stat_arb_portfolio(strategy_id: u64) -> Result<(), Error> {
    let strategy = get_stat_arb_strategy_mut(strategy_id)?;
    
    let portfolio = match &strategy.active_portfolio {
        Some(p) => p,
        None => return Ok(()),
    };
    
    // Check if rebalance needed
    let elapsed = current_time() - portfolio.entry_time;
    if elapsed < (strategy.rebalance_frequency_hours as u64 * 3600) {
        return Ok(());
    }
    
    // Recalculate optimal hedge ratios
    let coint_test = test_cointegration(
        strategy.asset_basket.clone(),
        strategy.lookback_period_days
    )?;
    
    if !coint_test.is_cointegrated {
        // No longer cointegrated - close position
        return close_stat_arb_portfolio(strategy_id);
    }
    
    // Calculate target positions with new hedge ratios
    let total_value = portfolio.total_value;
    let total_hedge_ratio: i128 = coint_test.hedge_ratios.iter().sum();
    
    for (i, asset) in strategy.asset_basket.iter().enumerate() {
        let target_weight = coint_test.hedge_ratios[i];
        let target_amount = (total_value * target_weight) / total_hedge_ratio;
        
        let current_amount = portfolio.asset_positions.get(asset).copied().unwrap_or(0);
        let delta = target_amount - current_amount;
        
        if delta.abs() > (target_amount * 50 / 10000) { // Rebalance if >0.5% deviation
            execute_trade(
                strategy.user,
                *asset,
                delta.abs(),
                if delta > 0 {
                    OrderType::Market // Buy more
                } else {
                    OrderType::Market // Sell some
                }
            )?;
        }
    }
    
    emit_event(StatArbPortfolioRebalanced {
        strategy_id,
        portfolio_id: portfolio.portfolio_id,
    });
    
    Ok(())
}
Exit Monitoring:
rustfn check_stat_arb_exit(strategy_id: u64) -> Result<Option<u64>, Error> {
    let strategy = get_stat_arb_strategy_mut(strategy_id)?;
    
    let portfolio = match &strategy.active_portfolio {
        Some(p) => p,
        None => return Ok(None),
    };
    
    // Calculate current portfolio residual
    let mut current_prices = Vec::new();
    let mut hedge_ratios = Vec::new();
    
    for asset in &strategy.asset_basket {
        current_prices.push(get_current_price(*asset)?);
        let position = portfolio.asset_positions.get(asset).copied().unwrap_or(0);
        hedge_ratios.push(position);
    }
    
    let current_residual = calculate_portfolio_value(&current_prices, &hedge_ratios)?;
    
    // Calculate z-score
    let mean = strategy.residual_history.iter().sum::<i128>() / 
               strategy.residual_history.len() as i128;
    let std_dev = calculate_std_dev(&strategy.residual_history)?;
    let z_score = ((current_residual - mean) * 10000) / std_dev;
    
    // Exit if converged to mean or stop loss
    let converged = z_score.abs() <= strategy.exit_z_score;
    let stop_loss = z_score.abs() >= (portfolio.entry_z_score.abs() * 2);
    
    if converged || stop_loss {
        close_stat_arb_portfolio(strategy_id)?;
        
        let reason = if converged { "Converged" } else { "Stop loss" };
        
        emit_event(StatArbPortfolioClosed {
            strategy_id,
            portfolio_id: portfolio.portfolio_id,
            exit_z_score: z_score,
            reason: reason.to_string(),
        });
        
        return Ok(Some(portfolio.portfolio_id));
    }
    
    Ok(None)
}

fn close_stat_arb_portfolio(strategy_id: u64) -> Result<(), Error> {
    let strategy = get_stat_arb_strategy_mut(strategy_id)?;
    
    let portfolio = match &strategy.active_portfolio {
        Some(p) => p,
        None => return Ok(()),
    };
    
    // Close all positions
    for (asset, amount) in &portfolio.asset_positions {
        execute_trade(
            strategy.user,
            *asset,
            amount.abs(),
            OrderType::Market
        )?;
    }
    
    strategy.active_portfolio = None;
    
    Ok(())
}
Edge Cases:

Cointegration breaks down during position (close immediately)
One asset delisted or halted (close entire portfolio)
Extreme divergence (risk of permanent breakdown - stop loss)
Insufficient liquidity for rebalancing (partial rebalance or close)

Definition of Done:

 Cointegration testing (ADF)
 Optimal hedge ratio calculation
 Multi-asset portfolio construction
 Mean reversion trading on residuals
 Portfolio rebalancing
 Half-life calculation
 Unit tests verify statistical methods

Validation:

Test 3 assets for cointegration
Verify ADF statistic < -3.4
Calculate hedge ratios
Construct portfolio
Simulate residual divergence
Execute stat arb trade
Monitor convergence
Rebalance portfolio
Close on convergence
